File: plover_vm/micro/cw.py
"""10-bit control word fields (lo = bus/ALU, hi = REG_SEL)."""

from __future__ import annotations

from dataclasses import dataclass

CW_FLASH_BASE = 0x4000


@dataclass(frozen=True)
class ControlWord:
    raw: int

    @property
    def lo(self) -> int:
        return self.raw & 0xFF

    @property
    def reg_sel(self) -> int:
        return (self.raw >> 8) & 3

    @property
    def alu_op(self) -> int:
        return (self.lo >> 4) & 0xF

    @property
    def reg_we(self) -> bool:
        return bool((self.lo >> 3) & 1)

    @property
    def y_oe(self) -> bool:
        return bool((self.lo >> 2) & 1)

    @property
    def mem_rd(self) -> bool:
        return bool((self.lo >> 1) & 1)

    @property
    def mem_wr(self) -> bool:
        return bool(self.lo & 1)


def cs_index(opcode: int, phase: int) -> int:
    return ((opcode & 0xF) << 2) | (phase & 3)


def lookup_cw(nor_read_fn, opcode: int, phase: int) -> ControlWord:
    idx = cs_index(opcode, phase)
    return ControlWord(nor_read_fn(idx))

File: crates/plover_core/src/micro/cw.rs
#[derive(Clone, Copy, Debug)]
pub struct ControlWord {
    pub raw: u16,
}

impl ControlWord {
    fn lo(self) -> u8 {
        (self.raw & 0xFF) as u8
    }

    pub fn reg_sel(self) -> u8 {
        ((self.raw >> 8) & 3) as u8
    }

    pub fn alu_op(self) -> u8 {
        (self.lo() >> 4) & 0xF
    }

    pub fn reg_we(self) -> bool {
        (self.lo() >> 3) & 1 != 0
    }

    pub fn y_oe(self) -> bool {
        (self.lo() >> 2) & 1 != 0
    }

    pub fn mem_rd(self) -> bool {
        (self.lo() >> 1) & 1 != 0
    }

    pub fn mem_wr(self) -> bool {
        self.lo() & 1 != 0
    }
}

pub fn cs_index(opcode: u8, phase: u8) -> usize {
    ((opcode & 0xF) as usize) << 2 | (phase & 3) as usize
}

pub fn lookup_cw(mut nor_read: impl FnMut(usize) -> u16, opcode: u8, phase: u8) -> ControlWord {
    let idx = cs_index(opcode, phase);
    ControlWord {
        raw: nor_read(idx),
    }
}
